File: pRobot169/perl/days_statistics.pl
#!/usr/bin/perl
#

my (%first, %last);
my (%authors, %titles);
my (%selfErase);
my ($articles, $deleted);

while (my $dir = shift) {
    next if ! -d $dir;
    my $date = "";
    # --- get date -----------------
    1 while $dir =~ s#/$##;
    if ($dir =~ m#.*/([^/]+)$#) { $date = $1; }
    else { $date = $dir; }
    $date = substr($date, 0, 8);
    # ==============================
    next if (! $date =~ /^\d{8}$/);
    next if ! -f "$dir/stat.txt";
    next if ! -f "$dir/$date.txt";

    my ($start, $last) = &getStartLast("$dir/stat.txt");
    my $slashdate = substr($date,4,2)."/".substr($date,6,2);
    open LIST, "<$dir/$date.txt";
    while (<LIST>) {
        next if (substr($_, 10, 5) ne $slashdate);
        next if (substr($_, 1, 5) lt $start);
        last if (substr($_, 1, 5) gt $last);
        chomp;
        s/\r//g; s/\n//g;
        my $author = substr($_, 16, 12);
        $author =~ s/ +$//;
        my $title = substr($_, 32);
        $first{$date} = $_ if (substr($_, 1, 5) eq $start);
        $last{$date} = $_ if (substr($_, 1, 5) eq $last);

        $authors{$author}++;
        if (lc(substr($_, 7, 1)) eq "d") {
            $deleted++;
            if ($title =~ /\($author\)/) {
                $selfErase{$author}++;
            }
            next;
        }
        $articles++;
        $titles{$title}++;
    }
    close LIST;

    print "$date $slashdate $start-$last\n";
}

#&printAuthors;
&printTitles;

sub topN ($$$) {
    my ($pList, $nTop, $nChar) = @_;
    my $result;

    my $n = 0;
    my $nLast = 0;
    foreach my $i (sort { $b cmp $a } @$pList) {
        $result .= "$i\n" if $n++ < $nTop;
        $nLast = substr($i, 1, $nChar) if $n == $nTop;
        if (substr($i, 1, $nChar) < $nLast) {
            last;
        }
        elsif ($n > $nTop) {
            $result .= "$i\n";
        }
    }
    return $result;
}

sub printTitles {
    foreach my $i (keys %titles) {
        push @sortTitles, sprintf("%6s | %s", $titles{$i}, $i);
    }

    my @titlesNormal;
    my @titlesPopular;

    foreach my $i (@sortTitles) {
        if ($i =~ /[0-9]/) {
            if ($i =~ /(推廣|日目|航目)/) {
                push @titlesPopular, $i;
            }
            else {
                push @titlesNormal, $i;
            }
        }
    }

    my @sumTitlesPop = ("111111 | \xFF\xFF");
    my %mainTitlesPop;
    for my $i (sort { substr($a, index($a,"|")+2) cmp
            substr($b, index($b,"|")+2) } @titlesPopular) {
        $sumTitlesPop[$#sumTitlesPop] =~ /^\s*(\d+)\s\|\s(.+)$/;
        my ($nPrev, $titlePrev) = ($1, $2);
        $i =~ /^\s*(\d+)\s\|\s(.+)$/;
        my ($nCur, $titleCur) = ($1, $2);
        my ($ox, $titleMain) = &cmpSimilar($titlePrev, $titleCur);
        if ($ox eq "O") {
            pop @sumTitlesPop;
            push @sumTitlesPop, sprintf("%6s | %s", $nPrev+$nCur, $titlePrev);
            $mainTitlesPop{$titlePrev} = $titleMain;
        }
        else {
            push @sumTitlesPop, $i;
        }
    }
    shift @sumTitlesPop;

    for my $i (@sumTitlesPop) {
        $i =~ /^\s*(\d+)\s\|\s(.+)$/;
        my ($nCur, $titleCur) = ($1, $2);
        if (defined $mainTitlesPop{$titleCur}) {
            $i = sprintf("%6s | %s", $nCur, $mainTitlesPop{$titleCur});
        }
    }

    # --- print normal titles -----------------------------------
    print <<EOF;
     # |   Titles
-------+---------------------------------------------
EOF
    print &topN(\@titlesNormal, 40, 6);
    print "\n";
    # ===========================================================

    # --- print popularized titles ------------------------
    print <<EOF;
推廣成果:
     # |   Titles
-------+---------------------------------------------
EOF
    print &topN(\@sumTitlesPop, 20, 6);
    print "\n";
    # =====================================================
    print "\n";

}

sub printAuthors {
    print <<EOF;
     # |  Authors
-------+---------------------------------------------
EOF
    foreach my $i (keys %authors) {
        my $name = $i;
        $name = substr($i,0,1).("*" x (length($i)-2)).substr($i,-1,1);
        substr($name,0,1) = "*" if length($name)==2;
        push @sortAuthors, sprintf("%6s | %1s", $authors{$i}, $name);
    }
    print &topN(\@sortAuthors, 30, 6);
    print "\n";
}

sub getStartLast ($) {
    my $stat = shift;
    my ($start, $last) = undef;

    open STAT, "<$stat";
    while (<STAT>) {
        $start = sprintf("%5d",$1) if /^start = (\d+)/;
        $last = sprintf("%5d",$1) if /^last = (\d+)/;
    }
    close STAT;
    return ($start,$last);
}

sub cmpSimilar ($$) {
    my $a = shift;
    my $b = shift;
    my ($maina, $mainb) = ($a, $b);
    my ($main1, $main2) = ($maina, $mainb);
    ($main1, $main2) = ($main2, $main1) if (length($main1)>length($main2));
    if (substr($main2,0,length($main1)) eq $main1
        && substr($main2,length($main1)) =~ /^[a-zA-Z ]{0,3}$/) {
        return ("O", $maina);
    }

    my @a = grep /./, split(/([\x80-\xFF].|.)/,$a);
    my @b = grep /./, split(/([\x80-\xFF].|.)/,$b);
    my $i = 0;
    sub isNumChi;
    for $i (0 .. $#a) {
        last if $i > $#b;
        if ($a[$i] ne $b[$i]) {
            if (&isNumChi($a[$i]) || &isNumChi($b[$i])) {
                my $ja = $i;
                my $jb = $i;
                my $ka = $i;
                my $kb = $i;
                --$ja if (defined $a[$i-1] && &isNumChi($a[$i-1])
                    && ! &isNumChi($a[$i]));
                --$jb if (defined $b[$i-1] && &isNumChi($b[$i-1])
                    && ! &isNumChi($b[$i]));
                --$ja while ($ja > 0 && &isNumChi($a[$ja]));
                --$jb while ($jb > 0 && &isNumChi($b[$jb]));
                if ($ja != $jb) {
                    return ("X", $maina);
                }
                ++$ka while ($ka<=$#a && &isNumChi($a[$ka]));
                ++$kb while ($kb<=$#b && &isNumChi($b[$kb]));
                my $maina = join("", @a[0 .. $ja], "---", @a[$ka .. $#a]);
                my $mainb = join("", @b[0 .. $jb], "---", @b[$kb .. $#b]);

                ($main1, $main2) = ($maina, $mainb);
                ($main1, $main2) = ($main2, $main1) if (length($main1)
                    >length($main2));
                if (substr($main2,0,length($main1)) eq $main1
                    && substr($main2,length($main1)) =~ /^[a-zA-Z ]{0,2}$/) {
                    return ("O", $maina);
                }
                else {
                    return ("X", $maina);
                }
                return;
            }
            elsif ($a[$i] =~ /\d/ || $b[$i] =~ /\d/) {
                my $ja = $i;
                my $jb = $i;
                my $ka = $i;
                my $kb = $i;
                --$ja if (defined $a[$i-1] && $a[$i-1] =~ /\d/
                    && ! ($a[$i] =~ /\d/));
                --$jb if (defined $b[$i-1] && $b[$i-1] =~ /\d/
                    && ! ($b[$i] =~ /\d/));
                --$ja while ($ja > 0 && $a[$ja] =~ /\d/);
                --$jb while ($jb > 0 && $b[$jb] =~ /\d/);
                if ($ja != $jb) {
                    return ("X", $maina);
                }
                ++$ka while ($ka<=$#a && $a[$ka] =~ /\d/);
                ++$kb while ($kb<=$#b && $b[$kb] =~ /\d/);
                my $maina = join("", @a[0 .. $ja], "---", @a[$ka .. $#a]);
                my $mainb = join("", @b[0 .. $jb], "---", @b[$kb .. $#b]);

                ($main1, $main2) = ($maina, $mainb);
                ($main1, $main2) = ($main2, $main1) if (length($main1)
                    >length($main2));
                if (substr($main2,0,length($main1)) eq $main1
                    && substr($main2,length($main1)) =~ /^[a-zA-Z ]{0,2}$/) {
                    return ("O", $maina);
                }
                else {
                    return ("X", $maina);
                }
                return;
            }
            else {
                return ("X", $a);
            }
        }
    }
}

sub isNumChi ($) {
    my $c = shift;
    if ($c eq '零' || $c eq '一' || $c eq '二' || $c eq '三'
        || $c eq '四' || $c eq '五' || $c eq '六' || $c eq '七'
        || $c eq '八' || $c eq '九' || $c eq '十' || $c eq '百'
        || $c eq '千' || $c eq '幾' ) {
        return 1;
    }
    return 0;
}

File: pRobot169/python/days_statistics.py
import os
import re

first = {}
last = {}
authors = {}
titles = {}
selfErase = {}
articles = 0
deleted = 0

def get_start_last(stat):
    start = last = None
    with open(stat, 'r') as stat_file:
        for line in stat_file:
            if match := re.match(r'^start = (\d+)', line):
                start = f"{int(match.group(1)):5}"
            if match := re.match(r'^last = (\d+)', line):
                last = f"{int(match.group(1)):5}"
    return start, last

def top_n(p_list, n_top, n_char):
    result = []
    n = 0
    n_last = 0
    for i in sorted(p_list, reverse=True):
        if n < n_top:
            result.append(i)
            n += 1
        n_last = int(i[1:n_char+1]) if n == n_top else n_last
        if int(i[1:n_char+1]) < n_last:
            break
        elif n > n_top:
            result.append(i)
    return '\n'.join(result)

def print_titles():
    global sort_titles
    sort_titles = [f"{titles[i]:6} | {i}" for i in titles]

    titles_normal = []
    titles_popular = []

    for i in sort_titles:
        if re.search(r'[0-9]', i):
            if re.search(r'(推廣|日目|航目)', i):
                titles_popular.append(i)
            else:
                titles_normal.append(i)

    sum_titles_pop = ["111111 | \xFF\xFF"]
    main_titles_pop = {}
    for i in sorted(titles_popular, key=lambda x: x.split("|")[1].strip()):
        n_prev, title_prev = re.match(r'^\s*(\d+)\s\|\s(.+)$', sum_titles_pop[-1]).groups()
        n_cur, title_cur = re.match(r'^\s*(\d+)\s\|\s(.+)$', i).groups()
        ox, title_main = cmp_similar(title_prev, title_cur)
        if ox == "O":
            sum_titles_pop.pop()
            sum_titles_pop.append(f"{int(n_prev)+int(n_cur):6} | {title_prev}")
            main_titles_pop[title_prev] = title_main
        else:
            sum_titles_pop.append(i)
    sum_titles_pop.pop(0)

    for i in sum_titles_pop:
        n_cur, title_cur = re.match(r'^\s*(\d+)\s\|\s(.+)$', i).groups()
        if title_cur in main_titles_pop:
            i = f"{int(n_cur):6} | {main_titles_pop[title_cur]}"

    print("     # |   Titles\n-------+---------------------------------------------")
    print(top_n(titles_normal, 40, 6))
    print("\n推廣成果:\n     # |   Titles\n-------+---------------------------------------------")
    print(top_n(sum_titles_pop, 20, 6))
    print("\n")

def print_authors():
    print("     # |  Authors\n-------+---------------------------------------------")
    global sort_authors
    sort_authors = []
    for i in authors:
        name = i
        name = i[0] + "*" * (len(i) - 2) + i[-1]
        if len(name) == 2:
            name = "*" + name[1]
        sort_authors.append(f"{authors[i]:6} | {name}")
    print(top_n(sort_authors, 30, 6))
    print("\n")

def cmp_similar(a, b):
    maina, mainb = a, b
    main1, main2 = maina, mainb
    if len(main1) > len(main2):
        main1, main2 = main2, main1
    if main2.startswith(main1) and re.match(r'^[a-zA-Z ]{0,3}$', main2[len(main1):]):
        return "O", maina

    a_chars = [char for char in a if char.strip()]
    b_chars = [char for char in b if char.strip()]
    i = 0
    for i in range(len(a_chars)):
        if i > len(b_chars) - 1:
            break
        if a_chars[i] != b_chars[i]:
            if is_num_chi(a_chars[i]) or is_num_chi(b_chars[i]):
                ja = i
                jb = i
                ka = i
                kb = i
                if ja > 0 and is_num_chi(a_chars[ja-1]) and not is_num_chi(a_chars[i]):
                    ja -= 1
                if jb > 0 and is_num_chi(b_chars[jb-1]) and not is_num_chi(b_chars[i]):
                    jb -= 1
                while ja > 0 and is_num_chi(a_chars[ja]):
                    ja -= 1
                while jb > 0 and is_num_chi(b_chars[jb]):
                    jb -= 1
                if ja != jb:
                    return "X", maina
                while ka <= len(a_chars) - 1 and is_num_chi(a_chars[ka]):
                    ka += 1
                while kb <= len(b_chars) - 1 and is_num_chi(b_chars[kb]):
                    kb += 1
                maina = ''.join(a_chars[:ja+1] + ["---"] + a_chars[ka:])
                mainb = ''.join(b_chars[:jb+1] + ["---"] + b_chars[kb:])
                main1, main2 = maina, mainb
                if len(main1) > len(main2):
                    main1, main2 = main2, main1
                if main2.startswith(main1) and re.match(r'^[a-zA-Z ]{0,2}$', main2[len(main1):]):
                    return "O", maina
                else:
                    return "X", maina
            elif re.match(r'\d', a_chars[i]) or re.match(r'\d', b_chars[i]):
                ja = i
                jb = i
                ka = i
                kb = i
                if ja > 0 and re.match(r'\d', a_chars[ja-1]) and not re.match(r'\d', a_chars[i]):
                    ja -= 1
                if jb > 0 and re.match(r'\d', b_chars[jb-1]) and not re.match(r'\d', b_chars[i]):
                    jb -= 1
                while ja > 0 and re.match(r'\d', a_chars[ja]):
                    ja -= 1
                while jb > 0 and re.match(r'\d', b_chars[jb]):
                    jb -= 1
                if ja != jb:
                    return "X", maina
                while ka <= len(a_chars) - 1 and re.match(r'\d', a_chars[ka]):
                    ka += 1
                while kb <= len(b_chars) - 1 and re.match(r'\d', b_chars[kb]):
                    kb += 1
                maina = ''.join(a_chars[:ja+1] + ["---"] + a_chars[ka:])
                mainb = ''.join(b_chars[:jb+1] + ["---"] + b_chars[kb:])
                main1, main2 = maina, mainb
                if len(main1) > len(main2):
                    main1, main2 = main2, main1
                if main2.startswith(main1) and re.match(r'^[a-zA-Z ]{0,2}$', main2[len(main1):]):
                    return "O", maina
                else:
                    return "X", maina
            else:
                return "X", a
    return "X", a

def is_num_chi(c):
    num_chi = {'零', '一', '二', '三', '四', '五', '六', '七', '八', '九', '十', '百', '千', '幾'}
    return c in num_chi

import sys
directories = sys.argv[1:]

for dir in directories:
    if not os.path.isdir(dir):
        continue
    date = ""
    while dir.endswith('/'):
        dir = dir[:-1]
    if match := re.match(r'.*/([^/]+)$', dir):
        date = match.group(1)
    else:
        date = dir
    date = date[:8]
    if not re.match(r'^\d{8}$', date):
        continue
    if not os.path.isfile(f"{dir}/stat.txt"):
        continue
    if not os.path.isfile(f"{dir}/{date}.txt"):
        continue

    start, last = get_start_last(f"{dir}/stat.txt")
    slashdate = f"{date[4:6]}/{date[6:8]}"
    with open(f"{dir}/{date}.txt", 'r') as list_file:
        for line in list_file:
            if line[10:15] != slashdate:
                continue
            if line[1:6] < start:
                continue
            if line[1:6] > last:
                break
            line = line.strip().replace('\r', '').replace('\n', '')
            author = line[16:28].rstrip()
            title = line[32:]
            if line[1:6] == start:
                first[date] = line
            if line[1:6] == last:
                last[date] = line

            authors[author] = authors.get(author, 0) + 1
            if line[7:8].lower() == "d":
                deleted += 1
                if re.search(fr'\({author}\)', title):
                    selfErase[author] = selfErase.get(author, 0) + 1
                continue
            articles += 1
            titles[title] = titles.get(title, 0) + 1

    print(f"{date} {slashdate} {start}-{last}\n")

print_titles()

